Add openrollup-mvp-phase-1 application

chrisicen opened the pull request:
### Project Abstract

Open rollup provides a powerful and complete zk-rollup solution for substrate-based chains.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying? 
- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $50,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] A BTC or Ethereum (DAI/USDT) address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledged the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).

### How Did You Hear About our grants program?

- [ ] Social Media
- [ ] Hackathon
- [x] Personal Recommendation
- [ ] Substrate Builders Program
- [ ] Investor/VC
- [ ] Online Search 
- [ ] Other: _______

Noc2 reviewed (changes requested):
Thanks a lot for the application and your interest in the grants program. Do you already have a whitepaper or something similar in which you describe your rollup solution in more detail? For example, how does it compare to other existing rollups? Which curves/zkps are you planning to use? Also is Cairo and zksync already fully open source? In any case, I would recommend focusing on only one solution, at least at the beginning.

chrisicen commented:
> Thanks a lot for the application and your interest in the grants program. Do you already have a whitepaper or something similar in which you describe your rollup solution in more detail? For example, how does it compare to other existing rollups? Which curves/zkps are you planning to use? Also is Cairo and zksync already fully open source? In any case, I would recommend focusing on only one solution, at least at the beginning.

Thanks a lot for the feedback.

We plan to use [winterfell](https://github.com/novifinancial/winterfell/) to generate proofs for Cairo programs and verify the execution of Cairo programs.

If we're not missing something, Cairo's STARK prover is not open source, instead, the [SHARP](https://www.cairo-lang.org/docs/sharp.html) service generates the proof. So we need a Cairo VM's prover and verifier in Rust, the main part is Cairo VM's AIR (Algebraic Intermediate Representation) of STARK.

STARK's prover and verifier do not involve elliptic curves. For zkapp (L2) authentication we will test secp256k1 and [STARK curve](https://docs.starkware.co/starkex-docs-v2-deprecated/crypto/stark-curve), we prefer to use secp256k1 as used in substrate.

We need more time to research and test the technology stack we use and the compatibility of winterfell and Cairo, so we hope to leave it for the next proposal.

One difference between Open rollup and other rollups is that it is more decentralized, that is, each zkapp is responsible for submitting its own L2 transaction.  And as all zkapps interact with the same Open rollup pallet, user assets can be moved directly from one zkapp to another. 

At present, the rollups project is mainly on Ethereum, and we hope that Open rollup can enrich the polkadot/substrate ecosystem.

Thanks again, feel free to let me know if I need to add anything more.

Noc2 reviewed (changes requested):
Sorry for the late reply here and thanks for the response. So just to double check the verification isn’t part of this proposal, correct? In this case, the proposal looks relatively expensive to me for a simple pallet and I would suggest adding additional features.

chrisicen commented:
> Sorry for the late reply here and thanks for the response. So just to double check the verification isn’t part of this proposal, correct? In this case, the proposal looks relatively expensive to me for a simple pallet and I would suggest adding additional features.

Hi, @Noc2 , Thanks for your reply, yes this proposal does not include verification.

If it is just a specific program (circuit or AIR) verifier, it is relatively simple, there are some open source implementations of Groth16 (e.g., [hertmez](https://github.com/hermeznetwork/contracts/blob/master/contracts/hermez/lib/verifier2048.sol)) and plonk (e.g., zksync) verifier.

What we want to implement is a general computation proof and verification system which support any application which generally requires the realization of a verifiable virtual machine that is verified fast than compute directly. for example [miden](https://github.com/maticnetwork/miden/tree/main/miden) (A STARK-based stack virtual machine.), [zkevm](https://github.com/appliedzkp/zkevm-specs), and cairo. This is far more complicated than verifying a single program, so we plan to implement rollup and verification system separately. In this proposal, only include a prover trait suitable for general apps and a mock prover.

Although this proposal does not include the verification, this proposal also includes a complete rollup implemented for general apps, which needs to handle all L1 and L2 transactions, user deposits, exits, processing batches submitted by apps (update states, compute batch commits, verify that the batch is correct, etc.). 

Unlike [zksync](https://github.com/matter-labs/zksync/blob/master/contracts/contracts/ZkSync.sol) and [hermez](https://github.com/hermeznetwork/contracts/blob/master/contracts/hermez/Hermez.sol)'s rollup that only supports their single application, Open rollup supports arbitrary developers to register their applications, so open rollup needs to store and process the states and data of multiple applications and supports user assets move between apps. 

We have revised our proposal to add support for asserts (tokens) and NFTs in Assets and Uniques pallets which are used by [statement parachain](https://github.com/paritytech/cumulus/blob/master/polkadot-parachains/statemint/src/lib.rs#L535), which make open rollup more complete and hope to make our price more reasonable.

Looking forward to your reply.

Noc2 reviewed (changes requested):
Thanks for the reply. However, the milestone and its deliverables ultimately define what you need to deliver as part of the grant and it currently only mentions more or less the pallet, which doesn’t seem to contain the following: 

>  In this proposal, only include a prover trait suitable for general apps and a mock verification.

> We have revised our proposal to add support for asserts (tokens) and NFTs of Assets and Uniques pallets which are used by [statement parachain](https://github.com/paritytech/cumulus/blob/master/polkadot-parachains/statemint/src/lib.rs#L535), which make open rollup more complete and hope to make our price more reasonable.

Could you update this?

chrisicen commented:
> Thanks for the reply. However, the milestone and its deliverables ultimately define what you need to deliver as part of the grant and it currently only mentions more or less the pallet, which doesn’t seem to contain the following:
> 
> > In this proposal, only include a prover trait suitable for general apps and a mock verification.
> 
> > We have revised our proposal to add support for asserts (tokens) and NFTs of Assets and Uniques pallets which are used by [statement parachain](https://github.com/paritytech/cumulus/blob/master/polkadot-parachains/statemint/src/lib.rs#L535), which make open rollup more complete and hope to make our price more reasonable.
> 
> Could you update this?

@Noc2 Thanks for the feedback, the deliverables have been updated.

Noc2 reviewed (changes requested):
Thanks for the update. You integrated a “verifier trait suitable for general apps”. Above you said something about a “prover trait suitable for general apps”, so was this on purpose? Also could you integrate more information about how this would work for general apps?

chrisicen commented:
Hi @Noc2 , Thanks for the feedback!

>  You integrated a “verifier trait suitable for general apps”. Above you said something about a “prover trait suitable for general apps”, so was this on purpose? 

:flushed: The "prover" is a typo, the on-chain pallet only needs verifiers.

> Also could you integrate more information about how this would work for general apps?

We've added Prove and Verification abstraction in the latest revision, look forward to hearing back!

Noc2 reviewed (changes requested):
Thanks again for the update. I guess what I meant above is also if you are planning, as part of this grant, also to deliver any of the off-chain tools? If so, especially if you develop them anyway, feel free to add them to the delivery.

chrisicen commented:
> Thanks again for the update. I guess what I meant above is also if you are planning, as part of this grant, also to deliver any of the off-chain tools? If so, especially if you develop them anyway, feel free to add them to the delivery.

Hi @Noc2 , Thanks for your feedback! As a practical off-chain scalability solution, we plan to provide developers with a verifiable programming language development kit in the future, it is a bit difficult for us to provide a part of it in this grant as it is closely related to the specific zkvm. We really appreciate your feedback, and we feel the need to provide a zkapp demo example, which would make Open rollup easier to understand. We've added a demo example delivery in the last revision, looking forward to your reply.

Noc2 reviewed (changes requested):
Sorry for the late reply and thanks for the update. I personally find the grant a little bit expensive and it would be nice to have more details, but I will already share it with the rest of the team and also ping someone else, who knows more about the tech.

mmagician commented:
In its current state, it looks like a pallet which updates account balances of arbitrary users when some specific account submits a valid state transition, which in itself doesn't really constitute a rollup, but simply a way to update tx's in an aggregated fashion - but without verification of the state transitions, I don't see much value in this. 

As a suggestion of how to adapt the scope; I guess it would be fine to even have the tx's submitted on-chain for now, incl. a valid proof, and adding a sub-deliverable for an off-chain verifier, which, given the program, state and zk proof, can verify whether a given state transition is valid (whether the result of the verifier makes it on-chain or not is irrelevant for the first version, IMO).

mmagician reviewed (changes requested):
In its current state, it looks like a pallet which updates account balances of arbitrary users when some specific account submits a valid state transition, which in itself doesn't really constitute a rollup, but simply a way to update tx's in an aggregated fashion - but without verification of the state transitions, I don't see much value in this. 

As a suggestion of how to adapt the scope; I guess it would be fine to even have the tx's submitted on-chain for now, incl. a valid proof, and adding a sub-deliverable for an off-chain verifier, which, given the program, state and zk proof, can verify whether a given state transition is valid (whether the result of the verifier makes it on-chain or not is irrelevant for the first version, IMO).

takahser commented on applications/openrollup-mvp-phase-1.md:
```diff
@@ -0,0 +1,179 @@
+
+### Team's experience
+
+Our team  have much experience in applied cryptography, blockchain research, and software development. 
```
```suggestion
Our team has a lot of experience in applied cryptography, blockchain research, and software development. 
```

takahser commented on applications/openrollup-mvp-phase-1.md:
```diff
@@ -0,0 +1,179 @@
+### Team's experience
+
+Our team  have much experience in applied cryptography, blockchain research, and software development. 
+Chris has been a software engineer for over many years in the gaming industry and has led the development of various multiplayer online games. Recently he develop a hybrid consensus of Progpow and AuthorityRound [https://github.com/chrisicen/openethereum-unity](https://github.com/chrisicen/openethereum-unity). He is currently devoted to the development of off-chain ZK scalability solution for blo ...
```
```suggestion
Chris has been a software engineer for many years in the gaming industry and has led the development of various multiplayer online games. Recently he develop a hybrid consensus of Progpow and AuthorityRound [https://github.com/chrisicen/openethereum-unity](https://github.com/chrisicen/openethereum-unity). He is currently devoted to the development of an off-chain ZK scalability solution for blockchain.
```

takahser commented on applications/openrollup-mvp-phase-1.md:
```diff
@@ -0,0 +1,179 @@
+| 0c. | Testing Guide | Core functions will be fully covered by unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
+| 0d. | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 0e. | Article | We will publish an **article**/workshop that explains what was done/achieved as part of the grant. |
+| 1.  | Setup | We provide a repository that forms the basis of our deliverables.|
```
I'd suggest removing this deliverable, as it doesn't describe what you'll deliver but in what form (git repo).

takahser replied:
✅

takahser commented on applications/openrollup-mvp-phase-1.md:
```diff
@@ -0,0 +1,179 @@
+| 0d. | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 0e. | Article | We will publish an **article**/workshop that explains what was done/achieved as part of the grant. |
+| 1.  | Setup | We provide a repository that forms the basis of our deliverables.|
+| 2.  | Core Types | We provide the core data types, as specified in Project Details. |
```
So these are the data structures, that will be used in your pallet? If so, I think it's sufficient to integrate it into the pallet deliverable. If there is a specific reason to do it in a separate fashion, e.g. in a dedicated library, pls elaborate on the reasons why you'd want to do it this way.

takahser replied:
✅

takahser commented on applications/openrollup-mvp-phase-1.md:
```diff
@@ -0,0 +1,179 @@
+| 0e. | Article | We will publish an **article**/workshop that explains what was done/achieved as part of the grant. |
+| 1.  | Setup | We provide a repository that forms the basis of our deliverables.|
+| 2.  | Core Types | We provide the core data types, as specified in Project Details. |
+| 3.  | Verifier trait and mock verifier | We provide the verifier trait suitable for general apps and a mock verifier that always pass the zk-verification. |
+| 4.  | Pallet functions | We provide the functionalities(zkapp registration/user deposit/user exit/user full exit/zkapp batch submit/zkapp management/zkapp info api) which support currencies/tokens/NFTs as specified in Project Details. |
```
This both sounds to me like you're describing the pallet itself. If that's the case, feel free to merge the two deliveries into 1.

chrisicen replied:
hi, @takahser , I would like our pallet to support various zkvms, and the recent revision added miden verifier support in this delivery,  so it's a little clearer to keep a separate delivery.

takahser reviewed (changes requested):
@chrisicen if I'm not mistaken, you're intending to implement 1 pallet here, is that correct? It seems like a rather high price to me for what you're planning to deliver, so feel free to reduce the price. Also I added a couple of comments and suggestions, feel free to have a look at them.
Please also have a look at the comments of the other team members @Noc2 and @mmagician and feel free to elaborate on the points they made.

chrisicen commented:
Hi @Noc2 @mmagician @takahser , thanks a lot for the feedback and @takahser's corrections of my English mistakes.

There are many zkvms, such as [cairo](https://www.cairo-lang.org/), [miden](https://github.com/maticnetwork/miden/tree/main/miden), [zkevm](https://github.com/privacy-scaling-explorations/zkevm-specs), but it is still too difficult for application developers. We will also propose our own scheme in the future, which is why we did not integrate any kind of zk proof system in our original proposal, because we want to spend more time in our own zkvm scheme.

After our consideration, we decided to add the miden proof system that is developed using rust and does not require rust's std standard library so that it can be used directly in the substrate pallet. It does not require much time and makes open rollup to be a complete L2 solution, we want to add support for miden to keep our original price.

For a zkvm to become a practical solution in the blockchain, a well-designed interaction protocol between the zkvm‘s applications and the chain is very important, we hope that our proposed open rollup can become an open protocol for off-chain scaling solutions, this enables a new zkvm scheme to be quickly used in the blockchain world.

Thanks again for your suggestions and patient responses, feel free to tell us what else we need to change.

Noc2 reviewed (changes requested):
Sorry for the very late reply here. Are you aware of the following proposal: https://github.com/w3f/Grants-Program/blob/master/applications/starks_network.md#development-roadmap-nut_and_bolt? Also to double check, so you plan implementing Polygon miden for substrate now, correct? Your delivery: Verifier trait and miden verifier are currently not very specific in this regard. Could you clarify this?

mmagician commented:
I agree there with @Noc2 , in that the deliverable of "Verifier trait and miden verifier" is too vague. It's not clear to me whether you want to make a substrate pallet, standalone library or something else entirely?

Also, have you already thought about whether all the required cryptographic primitives for STARKs are performant enough in a WASM? If I understand correctly, you're going to be running a (the Miden VM) in a WASM environment (i.e. in another VM) - and any dApp or program would only run in the top-level VM (miden). I'm not extremely familiar with STARKs or what primitives you'd need exactly, but for SNARK-based protocols deploying cryptographic libs straight-off in WASM is unaccaptably slow and sophisticated tricks are needed to actually pull it off (e.g. delegating the expensive calls to the substrate host).

mmagician commented:
Fair enough - in that case I guess it'd be intersting to see that implemented in substrate. We haven't had STARKs yet, AFAIK. 

Still, your deliverables are very minimalistic in their description. I suggest you provide as much details as possible (e.g. what interfaces are going to be implemented in substrate, what data you're gonna store in the pallet, etc.), so that grant evaluation upon delivery will be easier. Currently, there is too much leeway in the way they are specified, especially for the current price tag.

Noc2 reviewed (changes requested):
Thanks for the quick reply. As @mmagician pointed out could you please update the milestone deliverables and the specification.

chrisicen commented:
Hi @Noc2  @mmagician  Thanks for your feedback!

I have updated the milestone deliverables in the last revision, looking forward to your reply.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/w3f/Grants-Program?pullRequest=912) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/w3f/Grants-Program?pullRequest=912) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/w3f/Grants-Program?pullRequest=912) it.</sub>

Noc2 reviewed (changes requested):
Sorry for the late response here. I missed your reply and thanks for the update. I have two follow-up questions here: 

- Which pallet functions will make use of the WASM Miden verifier? 
- Would you be willing to reduce the price? As @takahser pointed out your application is currently rather expensive.

chrisicen commented:
>     * Which pallet functions will make use of the WASM Miden verifier?

This submitBatch function uses Miden verifier to verify the correctness of the Miden program execution.
I missed the proof parameter of the submitBatch function, I have added it.

>     * Would you be willing to reduce the price? As @takahser pointed out your application is currently rather expensive.

Thank you for your continued interest in the OpenRollup proposal, we have completed part of OpenRollup and we are happy to reduce the price. I have modified the price to 30K USD.

Noc2 reviewed (approved):
Thanks for the update. I’m happy to go ahead with it and mark it as ready for review.

cruikshankss reviewed (approved):
I think your "off-chain zkapp solution" has a lot of potential. Thank you for your application!

keeganquigley commented:
@chrisicen friendly reminder.

chrisicen commented:
> Hi team, how is M1 coming along? Is it possible to get a status update? Thx.

So sorry for not replying sooner, and sorry for not being able to delivery M1 on time. 

Recently, due to some other problems, the progress has been slow, and we should be able to complete the delivery by the end of February.

keeganquigley commented:
Thank @chrisicen for the update, that shouldn't be a problem. If you think it will take longer than that please open an [amendment](https://github.com/w3f/Grants-Program#changes-to-a-grant-after-approval) to extend the timeline, and that way there won't be any added pressure. Thanks!

chrisicen commented:
> Thank @chrisicen for the update, that shouldn't be a problem. If you think it will take longer than that please open an [amendment](https://github.com/w3f/Grants-Program#changes-to-a-grant-after-approval) to extend the timeline, and that way there won't be any added pressure. Thanks!

Thanks for the kind reminder, I just opened a pull request. #1500

keeganquigley commented:
Hi @chrisicen how is M1 coming along? Any updates you can give us?

keeganquigley commented:
Hi @chrisicen pinging for a friendly reminder.

chrisicen commented:
Hi @keeganquigley  very sorry for the late response and for not completing this proposal on time.

I underestimated how difficult this proposal would be for me. It took too much time to understand the type system of substrate, integrate nonfungible token and Miden-vm.

Currently, we have all the features completed and should be able to send delivery requests right away.

The code is in this project:  https://github.com/open-rollup

keeganquigley commented:
Thanks for the update @chrisicen. Do you still plan on submitting a delivery soon?

chrisicen commented:
Hi @keeganquigley I submitted the delivery yesterday. https://github.com/w3f/Grant-Milestone-Delivery/pull/843
